crates/gfx/src/engine.rs

```diff
@@ -100,94 +100,86 @@ pub enum MainMessage {
     SimulationError(GenericError),
 }
 
-/// An internal message passed to the simulation thread at engine-level.
+/// Rendering of the user interface.
+pub trait UIRenderer {
+    /// Set up the user interface for the given context.
+    fn render(context: &egui::Context);
+}
+
+/// Disables UI rendering.
 #[derive(Debug)]
-pub enum SimulationMessage {
-    /// Signals that the thread should gracefully shut down and exit.
-    Exit,
+pub struct NoUI;
+impl UIRenderer for NoUI {
+    fn render(_: &egui::Context) {}
 }
 
-impl<GfxInitFn, UIFn, SimulationFn, ClientContext, RenderData>
-    Engine<GfxInitFn, UIFn, SimulationFn, ClientContext, RenderData>
+/// A way of communicating (both sending data to and receiving data from) with the graphics engine.
+#[derive(Debug)]
+pub struct EngineHandle<RenderData> {
+    /// How the graphics engine receives information about what to render and where to render it.
+    pub render_data_sender: RingSender<RenderData>,
+    /// A receiver whose sender is dropped when the user has exited the application window.
+    ///
+    /// todo(#63): Change this so that the user code is responsible for handling the
+    /// todo(#63): "shutdown event" and can choose to drop the shutdown sender themselves.'
+    pub shutdown_receiver: Receiver<()>,
+    /// Can be used by the simulation thread to inform the main thread about events.
+    pub main_thread_sender: Sender<MainMessage>,
+}
+
+impl<UI, RenderData> Engine<UI, RenderData>
 where
-    for<'a> UIFn: Fn(&egui::Context) + 'a,
+    UI: UIRenderer + 'static,
     for<'a> RenderData: IntoIterator<Item = Object> + Send + 'a,
-    for<'a> ClientContext: Send + 'a,
-    for<'a> GfxInitFn: Fn(&mut ClientContext, &mut dyn Creator) -> GenericResult<()> + 'a,
-    for<'a> SimulationFn: FnMut(&mut ClientContext, &UpdateRate, &mut RingSender<RenderData>) -> GenericResult<()>
-        + Send
-        + Sync
-        + 'a,
 {
     /// Creates a new instance of `Engine`.
     ///
-    /// The `initialize_gfx` function is called once, before beginning the render loop.
-    /// The `simulate` function is called from the simulation thread every simulation tick.
+    /// [`GraphicsInitializer`] is called once, before beginning the render loop.
+    /// [`Simulation`] is called from the simulation thread every simulation tick.
```

Comment:
Cannot find `GraphicsInitializer` and `Simulation` in the code.

Reply:
Oops, that's old stuff. We'll remove it.